feat: initial Sway support

alterNERDtive opened the pull request:
Adds support for Sway; at least for freezing the _active window_. AFAIK there’s no utility like `hyprprop` for Sway. The detection logic relies on `loginctl` which is part of Systemd. If you’re not OK with that dependency, I’ll have to go an uglier route :)

Will output an error message if a command is run under a window manager that does not support it. As a side effect will also produce error messages if you are running it under a completely different window manager. WM-agnostig commands (like freezing by name or by PID) will still work anywhere.

Currently it will die if no information can be obtained from `loginctl`. If you prefer to silently assume Wayland/Hyprland, that’s trivial to implement.

Session/Desktop information added to `printInfo` as well.

Can you double check the output of `loginctl show-session $(loginctl | grep $(whoami) | awk '{print $1}') -p Type -p Desktop` on Hyprland for me, please? Also on any other DE/WM if you happen to have easy access to more of them, for reference.

Zerodya commented:
`loginctl show-session $(loginctl | grep $(whoami) | awk '{print $1}') -p Type -p Desktop` outputs 
```
Desktop=
Type=wayland
```
on NixOS Hyprland.

That said, I would rather not rely on Systemd at all as there are distros that use different init systems like Artix and Void (I used to be an Artix user myself and I know the pain), so a different approach would be better.

Zerodya commented on README.md:
```diff
@@ -1,5 +1,5 @@
 # hyprfreeze
-Hyprfreeze is a utility to suspend a game process (and other programs) in Hyprland.
+Hyprfreeze is a utility to suspend a game process (and other programs) on Wayland.
```
"Hyprland and Sway" would be better instead of "Wayland" because there are still a lot more wayland compositors that are not supported

alterNERDtive replied:
I was going to keep it open (in case someone wants to provide support for other desktops). I guess it’s simple enough to change again if that happens :)

alterNERDtive commented:
> `loginctl show-session $(loginctl | grep $(whoami) | awk '{print $1}') -p Type -p Desktop` outputs
> 
> ```
> Desktop=
> Type=wayland
> ```
> 
> on NixOS Hyprland.

That is … unexpected output. I’ll look into which other way I want to do then. Or probably even have multiple as fallbacks.

Does your NixOS set `XDG_SESSION_DESKTOP` / `XDK_SESSION_TYPE`

> That said, I would rather not rely on Systemd at all as there are distros that use different init systems like Artix and Void (I used to be an Artix user myself and I know the pain), so a different approach would be better.

Gotcha. If I have to implement multiple ways I can just check if `loginctl` is present before attempting to get data that way.

Zerodya commented:
> Does your NixOS set XDG_SESSION_DESKTOP / XDK_SESSION_TYPE

Yes
```bash
export | grep XDG_SESSION

XDG_SESSION_CLASS user
XDG_SESSION_DESKTOP hyprland
XDG_SESSION_ID 2
XDG_SESSION_TYPE wayland
```

Zerodya commented:
> Gotcha. If I have to implement multiple ways I can just check if loginctl is present before attempting to get data that way.

If you can find a way that works in Sway without systemd we should just use that. Let's not add more complexity and keep it systemd agnostic. Unless I misunderstood what you meant?

alterNERDtive commented:
OK, here’s my suggestion for round 2:

1. check if `loginctl` is present
2. if yes, use the current method
3. if not or if the values are not set, look at `$XDG_SESSION_*`
4. if neither of those worked, die

I’ll take a crack at it later. First, chores :<

AFAIK the `XDG` variables are not set correctly if you start your WM/DE without a display manager, straight from the tty; in which case they will just say `tty`.

Do you have any other ideas how to automatically detect the running desktop?

Zerodya commented:
Btw the default behaviour should be setting the desktop variable to Hyprland, something like this:
```
 DESKTOP = Hyprland
 
 if detectEnvironment() == Sway
   DESKTOP = Sway
```

alterNERDtive commented:
I think I’m basically done with this one now. I have changed “Wayland” back to “Hyprland/Sway”, removed the hard dependency on Systemd, added `$XDG_SESSION_DESKTOP` to the detection logic and added the fallback to Hyprland.

Zerodya commented:
Environment information should not be printed in `printInfo()` because that should only print information about the PID. It should be printed with `debugPrint` in `detectEnvironment()` when the desktop is successfully detected instead.

alterNERDtive commented:
> Environment information should not be printed in `printInfo()` because that should only print information about the PID. It should be printed with `debugPrint` in `detectEnvironment()` when the desktop is successfully detected instead.

roger that

Zerodya commented:
LGTM, thank you for the PR
